File: 1_Programming_and_Coding_Overview/Examples_Source/C/error_check.c
//------------------------------------------------------------------------------
// Name:        error_check.c
// Purpose:     Example
// Title:       "Input Validation Checks"
//
// Platform:    Win64, Ubuntu64
//
// Author:      Axle
// Created:     03/02/2022
// Updated:     19/02/2022
// Copyright:   (c) Axle 2022
// Licence:     MIT No Attribution
//------------------------------------------------------------------------------

// C std library headers.
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

// Helper functions
void S_Pause(void);
int S_getchar(void);
// v Emulates Input function in FreeBASIC and Python 3.
char *Input(char *buf, char *str, int n);
char *S_fgets(char *buf, int n, FILE *stream);
// Maximum buffer size (Max length of a string + 1 for '\0')
#define MAX_BUFFER 128  // Set a maximum size for buffers and never exceed it.

int main(void)  // Main procedure
    {
    char Input_Buffer[MAX_BUFFER] = {'\0'};  // buffer
    unsigned int Zero_Passlength = 0;  // No Password entered
    unsigned int Min_Passlength = 6;  // Pass length Min limit
    unsigned int Max_Passlength = 12;  // Pass length Max limit
    unsigned int Max_Attempts = 3;  // Attempts limit
    unsigned int Attempts_Counter = 0;  // Attempts count
    unsigned int Opt_Out = 0;  // Opt out question y/n

    // Test that the input data is within the expected limits of 6 to 12.
    // Loops until the user enters the correct data. In real life we would
    // also need an opt out after so many tries.
    // This may appear like a lot of extra code, but it is necessary in C for
    // safety. Unlike many "High Level" languages C does not have built in
    // safeguards so it is up to the coder to be pandantic about safety.
    // If you remove all of my "Comment explanations" you will find it is not
    // that much extra code :)
    // Note. C++, BASIC and Python have far more built in buffer safeguards,
    // but it is still up to the coder to ensure that any external input is
    // within the expected range of data.
    while(strlen(Input_Buffer) < Min_Passlength
        || strlen(Input_Buffer) > Max_Passlength)
        {
        // The next conditional check is part of an unwind to break out of deeply
        // nested loops. Although there are other methods to do this as well, I 
        // wanted to keep a simple method arcoss all 3 languages.
        // The above "While" conditional test should break the loop if the
        // Password is the correct length without the folowing test, but I wanted
        // to show an unwind method just the same :)
        if(Opt_Out == 'Y' || Opt_Out == 'y')
            {
            break;  // Quit asking for a name and break out of the loop "Step 2".
            // If we wanted to reuse the Opt_Out variable again later I suggest
            // uncommenting the following line.
            //Opt_Out = 0;  // Reset Y/N the response variable.
            }
        else
            {
            // Note that we have 2 safety checks for the string that can be entered.
            // the first is the MAX_BUFFER which limits all input to be less
            // than the buffer "n" Input( , , int n). this blocks the possability
            // of a buffer overflow if the user inputs a longer string than the
            // buffer can hold. See the comments in Input().
            Input(Input_Buffer, "Please enter your password.\n Between 6 to 12 letters:"
            , MAX_BUFFER -1);
            // The second part of our test is to see if the data is within the
            // range that is expected.
            if(strlen(Input_Buffer) == Zero_Passlength)  // 0 length string.
                {
                printf("You did not enter your password...\n");
                Attempts_Counter++;
                }
            else if((strlen(Input_Buffer) > Zero_Passlength)
                 && (strlen(Input_Buffer) < Min_Passlength))  // String shorter than 6.
                {
                printf("The password you entered is too short...\n");
                Attempts_Counter++;
                }
            else if(strlen(Input_Buffer) > Max_Passlength)  // String longer than 12.
                {
                printf("The password you entered is too long...\n");
                Attempts_Counter++;
                }
            else  // String is withing range. Success.
                {
                printf("Your Password is %s\n", Input_Buffer);
                }

            // Limit the number of attempts and offer an opt out so that the
            // user is not caught in an endless loop if they decide to not
            // enter a name.
            if(Attempts_Counter >= Max_Attempts)
                {
                // keep asking in the loop untill we get a valid Y/N response.
                while(Opt_Out != 'y' && Opt_Out != 'Y')
                    {
                    printf("\nSorry you have reached the maximum number of tries!\n");
                    printf("Would you like to quit? (Y/N):\n");
                    // Characters are unsigned integers in C: int Ch = 'Y' = (char)89;
                    // Strings are an array of characters char St = "Y"; = pionter
                    // to St[0] = (char)89 = (char)'Y'
                    Opt_Out =  S_getchar();
                    if(Opt_Out =='y' || Opt_Out == 'Y')
                        {
                        break;  // Quit asking for a name and break out "step 1".
                        }
                    else if(Opt_Out == 'n' || Opt_Out == 'N')
                        {
                        Attempts_Counter = 0;  // reset the attempts counter
                        Opt_Out = 0;  // reset the opt out counter
                        break;
                        }
                    else
                        {
                        // ask again until we get a Y/N response.
                        printf("Invalid response!\n");
                        }
                    }
                }
            }

        }

    printf("\nFile read error test\n");
    // Checking the error return of a function. This is always recomended when
    // the function handles data from an unknown source aka anything outside
    // of the source code of you application. This icludes "User Inputs",
    // "Data from a file or database", "Information from the web",
    // "Communication and data transfers to other apps" ++.
    // We can never guarentee the existance of data outside of our application
    // or if it will be the data that we have expected.
    //
    // Description of:
    // FILE *fopen(const char *filename, const char *mode)
    // Description
    // The C library function FILE *fopen(const char *filename, const char
    // *mode) opens the filename pointed to, by filename using the given mode.
    // Parameters:
    // filename -- This is the C string containing the name of the file to be
    // opened.
    // mode -- This is the C string containing a file access mode. It includes:
    // ...
    // Return Value: (<- this is what you need to look for.)
    // This function returns a FILE pointer. Otherwise, NULL is returned and the
    // global variable errno is set to indicate the error.
    // If it fails to open because the file does not exist for example, we need
    // to handle the error and either create the file, or tell the user that the
    // file could not be found, or any other number of options that are apropriate
    // to the context of your application. Don't ever let and error be passed to
    // your user with the horrible "Ding Sound" and the
    // "This program has terminated unexpectedly!" warning in production code. :)
    char *filename = "filename.txt";  // a dummy file name.
    FILE *fp;
    fp = fopen(filename, "r");
    // in this example "filename.txt" does not exist so the variable fp will be
    // sent an error as the return "NULL". (Null pointer is a special form of '0').
    // Yes it sounds weird, but computers have many types of Zero and not all
    // are equal NULL != null != 0 != -0 != FALSE; but are similar.
    if(fp == NULL)
        {
        // Please note that attempting to display a FILE* as an int %d is
        // classed as "undifined behavior" and will throw a compiler warning. You
        // could also use %p for void pointer, but that too is undefined.
        // I have only done this to show that it will display 0 but should not
        // be used or relied upon in production code. Always compare FILE* to
        // NULL. NULL is a special MACRO that represents the Zero value, or the
        // value of memory of a location of FILE * (pointer) to a structure that
        // does not exist.
        // if(fp == NULL) <- correct | if(fp == 0) <- undefined behavior.
        printf("ERROR! Cannot open file %s\n", filename);
        printf("fp * = %d\n", fp);
        // perror() will retreive the error sent to the console "stderr" and
        // print the value to "stdout".
        perror("Error in opening file");
        // Do some error handling tasks to deal with why the file does not exist.
        // Maybe you need to create the file first?
        // If this is a function that you have created you may wish to "return"
        // some usefull information to the function call.
        //return -1;  // -1 Indicates an error on some platforms.
        }
    else
        {
        // No errors, so do some file read operations.
        }
    fclose(fp);  // Always close files when finished, always.

    // It is common in C to use getchar() as a pause to wait for keybord [Enter]
    // to continue. If more than 1 character is entered then the remaining
    // characters are left in the keyboard buffer and appear next time the
    // keyboard buffer is read. including the new line '\n' for enter.
    // To eleiminate this we have to clear the keyboard buffer before or after
    // each entry. My function S_getchar() "Safe Get Character" enumerates all
    // characters in the buffer clearing any unused keyboard scan codes. The
    // next tim we call for a keyboard input there are no leftover artifacts in
    // the buffer :)
    // A another alternative than this is my Input() or S_fgets()
    // "Safe Get String"  based upon getc() "Get Char" which allows us to enter
    // a maximum nuber of characters to be retreived from stdin to our buffer. 
    // S_getchar() returns a single char as int, whereas S_fgets() "Safe Get String"
    // returns a char array (String) of maximum length n.
    S_Pause();
    return 0;
    }  // END main() <---

// --> START User defined functions and wrappers

// Safe Pause. A "Pause" wrapper for S_getchar().
void S_Pause(void)
    {
    // This function is referred to as a wrapper for S_getchar()
    printf("\nPress [Enter] to continue...");
    S_getchar();  // Uses S_getchar() for safety.
    }

// Safe getcar() removes all artefacts from the stdin buffer. This function
// is how we must always retreive keyboard single character input using a loop
// to enumerate all characters in the stdin buffer keeping only the first
// character and discarding any other characters including the '\n' [Enter].
// This finction is a wrapper that reduces the boiler plate code to a single
// line so we dont have to write the same loop every time we ask for a character
// from the keyboard. Modular function libraries are the key to convenient
// coding in C. C++ (High level and OOP) includes many of these libraries by
// default. C++ is a more convenient and safer upgrade to writting in C.
int S_getchar(void)
    {
    // This function is referred to as a wrapper for getchar()
    int i = 0;
    int ret;
    int ch;
    // The following enumerates all characters in the buffer.
    while((ch = getchar()) != '\n' && ch != EOF )
        {
        // But only keeps and returns the first char.
        if (i < 1)
            {
            ret = ch;
            }
        i++;
        }
    return ret;
    }

// Console Input (stdin).
// This a a wrapper function that emulates the "Input()" function found
// in BASIC and Python 3.
// buf is the return buffer, str is printed to the screen.
// returns a pointer to buf, str is the input message...
// n is the MAX number of characters allowed. This is considered a safe input
// function as it limits the length of the input string to n reducing the risk
// of a buffer overflow.
// buf must be at least n + 1 for '\0'
char *Input(char *buf, char *str, int n)
    {
    FILE *stream = stdin;
    char *empty = "";
    int ret;
    // Test if an empty string has been sent. If so...
    ret = strcmp(str, empty);
    if(ret != 0)  // Don't print an empty string.
        {
        printf("%s", str);  // Otherwise, Print the Input Message.
        }
    return S_fgets(buf, n, stream);  // Call my safe get string function.
    // and passes the truncated string back to the original function call.
    }

// Safe fgets() removes all artefacts from the stdin buffer.
// buf must be at least n + 1 for '\0'
// Returns a pointer to char array (String) of length n.
char *S_fgets(char *buf, int n, FILE *stream)
    {
    int i = 0;
    int ch;
    //memset(buf, 0, n);
    // The following enumerates all characters in the keybord buffer.
    // The while loop will exit if it encounters an [Enter] '\n' or
    // the end of the file/stream EOF.
    while((ch = getc(stream)) != '\n' && ch != EOF )
        {
        // But only keeps and returns n chars.
        // This is my safe string feature that only returns n charaters length
        // in the returned string. No buffer overflows for me :)
        // All additional charcters entered by the user are discarded.
        // It should be noted that the coder should promt the user to
        // "Enter a MAX of n characters".
        if (i < n)
            {
            buf[i] = ch;
            }
        i++;
        }
    buf[i] = '\0';
    return buf;  // returns the truncated string.
    }

File: 1_Programming_and_Coding_Overview/Examples_Source/Python_3/error_check.py
#------------------------------------------------------------------------------
# Name:        error_check.c
# Purpose:     Example
#
# Platform:    Win64, Ubuntu64
#
# Author:      Axle
# Created:     16/12/2021
# Updated:     18/02/2022
# Copyright:   (c) Axle 2021
# Licence:     MIT No Attribution
#------------------------------------------------------------------------------

def main():

    # The MAX size of a String type in Py is limited to the "Dynamic Memory"
    # available to the system. Pythons internals will not aloww the contents
    # of the input buffer to go above what is available to the system.
    # We can't set a static size for the length (Memory) in Python.
    Input_Buffer = ""
    Zero_Passlength = 0  # No Password entered
    Min_Passlength = 6  # Pass Min length limit
    Max_Passlength = 12  # Pass Max length limit
    Max_Attempts = 3  # Attempts limit
    Attempts_Counter = 0  # Attempts count
    Opt_Out = ""  # Opt out question y/n

    # Test that the input data is within the expected limits of 6 to 12.
    # Loops until the user enters the correct data. In real life we would
    # also need an opt out after so many tries.
    # This may appear like a lot of extra code, but it is necessary to keep a
    # level of safety.
    # If you remove all of my "Comment explanations" you will find it is not
    # that much extra code :)
    # C++, BASIC and Python have far more built in buffer safeguards than C,
    # but it is still up to the coder to ensure that any external input is
    # within the expected range of data.
    while ((len(Input_Buffer) < Min_Passlength) \
        or (len(Input_Buffer) > Max_Passlength)):
        # The next conditional check is part of an unwind to break out of deeply
        # nested loops. Although there are other methods to do this as well, I 
        # wanted to keep a simple method arcoss all 3 languages.
        # The above "While" conditional test should break the loop if the Password
        # is the correct length without the folowing test, but I wanted to show
        # an unwind method just the same :)
        if (Opt_Out == "Y") or (Opt_Out == "y"):
            break  # Quit asking for a name and break out of the loop "Step 2".
            # If we wanted to reuse the Opt_Out variable again later I suggest
            # uncommenting the following line.
            #Opt_Out = ""  ' Reset Y/N the response variable.
        else:
            # Note that Python has built in limit for the input string length
            # so we only have to check the the return is withing the range and
            # data type expected.
            print("Please enter your password.")
            Input_Buffer = input(" Between 6 to 12 letters:")
            # The second part of our test is to see if the data is within the
            # range that is expected.
            if (len(Input_Buffer) == Zero_Passlength): # 0 length string.
                print("You did not enter your password...")
                Attempts_Counter += 1
            elif ((len(Input_Buffer) > Zero_Passlength) \
                and (len(Input_Buffer) < Min_Passlength)): # String shorter than 6.
                print("The password you entered is too short...")
                Attempts_Counter += 1
            elif (len(Input_Buffer) > Max_Passlength):  # String longer than 12.
                print("The password you entered is too long...")
                Attempts_Counter += 1
            else:  # String is withing range. Success.
                print("Your Password is " + Input_Buffer)

            # Limit the number of attempts and offer an opt out so that the
            # user is not caught in an endless loop if they decide to not
            # enter a name.
            if (Attempts_Counter >= Max_Attempts):
                # keep asking in the loop untill we get a valid Y/N response.
                while (Opt_Out != "y") and (Opt_Out != "Y"):
                    print("\nSorry you have reached the maximum number of tries!")
                    Opt_Out = input("Would you like to quit? (Y/N):")
                    if (Opt_Out == "y") or (Opt_Out == "Y"):
                        break  # Quit asking for a name and break out "step 1".
                               # Opt_Out = "Y" will be used in step 2.
                    elif (Opt_Out == "n") or (Opt_Out == "N"):
                        # reset the attempts counter (3 more tries).
                        Attempts_Counter = 0 
                        Opt_Out = ""  # reset the opt out counter.
                        break
                    else:
                        # ask again until we get a Y/N response.
                        print("Invalid response!")

    print("")
    print("File read error test")
    # Checking the error return of a function. This is always recomended when
    # the function handles data from an unknown source aka anything outside
    # of the source code of you application. This icludes "User Inputs",
    # "Data from a file or database", "Information from the web",
    # "Communication and data transfers to other apps" ++.
    # We can never guarentee the existance of data outside of our application
    # or if it will be the data that we have expected.
    #
    # Description of:
    # open(file, mode='r', buffering=-1, encoding=None, errors=None, \
    # newline=None, closefd=True, opener=None)
    # Description
    # https://docs.python.org/3/library/functions.html#open
    # Return Value:
    # Python 3 only returns an "Object" and no error values like C and FreeBASIC.
    # If module core throws an error it is called an exception and the application
    # will terminate if the exception has not been handled. For this we use the
    # try: except method to "Catch" the exception and make a decission on how
    # to manage the error.
    # https://www.w3schools.com/python/python_try_except.asp
    # https://docs.python.org/3/tutorial/errors.html
    # https://docs.python.org/3/library/exceptions.html
    # If it fails to open because the file does not exist for example, we need
    # to handle the error and either create the file, or tell the user that the
    # file could not be found, or any other number of options that are apropriate
    # to the context of your application. Don't ever let and error be passed to
    # your user with the horrible "Ding Sound" and the
    # "This program has terminated unexpectedly!" warning in production code. :)
    # in this example "filename.txt" does not exist so open( ...) will
    # create an exception error FileNotFoundError(2, 'No such file or directory')
    # It is possible that the file may not yet exist. Opening it
    # as "r" will return an exception.
    filename = "filename.txt"  # a dummy file name.
    try:
        # NOTE! With open will automaticaly close the file handle fp so we
        # don't need to use close(fp).
        with open(filename, "r") as fp:  # Open the File.
            # No errors, so do some file read operations.
            Con_Pause()  # wait until a key is pressed

    except FileNotFoundError as e:  # Handle the exception error.
        print("\nERROR! Cannot open file " + filename)
        print("Maybe the file has not yet been created.")
        # 5 differnt way of retreiving the error message.
        # With e.args[0] we can retreive the error number for some functions.
        # it is not possible to list all the exception types and errors for
        # Python 3 so we usually attempt to cover the most common.
        # except Exception as e: # Will polulate e with any exception class
        # caught with except. You will need to break down the sublevels of each
        # child class to show the actual error such as "FileNotFoundError"
        print(e)
        print(e.args)
        print(e.args[0], e.args[1])
        print(repr(e))
        print(f"{type(e).__name__} at line {e.__traceback__.tb_lineno} of {__file__}: {e}")
        Con_Pause()

    Con_Pause()  # DEBUG Pause
    return None
    # END Main() <---

# Console Pause wrapper.
# This is just a simple wrapper for Input(). It is
# benificial over time to create your own personal function library for common
# task as it removes the need to type in the 3 lines of code below every time
# you need a pause. If you look at my Input() function from the C example you
# can see that I have reduced some 15 lines of code to a single function call.
# This forms the basic principles for the creation of "Modular Code" and code
# libraries. We can use a library function instead of repeadedly writting
# common ("Boiler Plate") code in our main() source.
def Con_Pause():
    dummy = ""
    print("")
    dummy = input("Press [Enter] key to continue...")
    return None

if __name__ == '__main__':
    main()
    